ruby/lib/google/protobuf/ffi/descriptor.rb

```diff
@@ -0,0 +1,765 @@
+            raise RuntimeError.new "Cannot parse a wrapper directly" if descriptor.send(:wrapper?)
+
+            if options[:ignore_unknown_fields]
+              decoding_options |= Google::Protobuf::FFI::Upb_JsonDecode_IgnoreUnknown
+            end
+
+            message = new
+            pool_def = pool_def_from_message_definition(message.class.descriptor)
+            status = Google::Protobuf::FFI::Status.new
+            unless Google::Protobuf::FFI.json_decode_message(data, data.bytesize, message.instance_variable_get(:@msg), message.class.descriptor, pool_def, decoding_options, message.instance_variable_get(:@arena), status)
+              raise ParseError.new "Error occurred during parsing: #{Google::Protobuf::FFI.error_message(status)}"
+            end
+            message
+          end
+
+          def self.encode_json(message, options = {})
+            encoding_options = 0
+            unless options.is_a? Hash
+              if options.respond_to? :to_h
+                options = options.to_h
+              else
+                #TODO(jatl) can this error message be improve to include what was received?
+                raise ArgumentError.new "Expected hash arguments"
+              end
+            end
+
+            if options[:preserve_proto_fieldnames]
+              encoding_options |= Google::Protobuf::FFI::Upb_JsonEncode_UseProtoNames
+            end
+            if options[:emit_defaults]
+              encoding_options |= Google::Protobuf::FFI::Upb_JsonEncode_EmitDefaults
+            end
+
+            buffer_size = 1024
+            buffer = ::FFI::MemoryPointer.new(:char, buffer_size)
+            status = Google::Protobuf::FFI::Status.new
+            msg = message.instance_variable_get(:@msg)
+            pool_def = pool_def_from_message_definition(message.class.descriptor)
+            size = Google::Protobuf::FFI::json_encode_message(msg, message.class.descriptor, pool_def, encoding_options, buffer, buffer_size, status)
+            unless status[:ok]
+              raise ParseError.new "Error occurred during encoding: #{Google::Protobuf::FFI.error_message(status)}"
+            end
+
+            if size >= buffer_size
+              buffer_size = size + 1
+              buffer = ::FFI::MemoryPointer.new(:char, buffer_size)
+              status.clear
+              size = Google::Protobuf::FFI::json_encode_message(msg, message.class.descriptor, pool_def, encoding_options, buffer, buffer_size, status)
+              unless status[:ok]
+                raise ParseError.new "Error occurred during encoding: #{Google::Protobuf::FFI.error_message(status)}"
+              end
+              if size >= buffer_size
+                raise ParseError.new "Inconsistent JSON encoding sizes - was #{buffer_size - 1}, now #{size}"
+              end
+            end
+
+            buffer.read_string_length(size).force_encoding("UTF-8").freeze
+          end
+
+          @descriptor.each do |field_descriptor|
```

Comment:
Ditto, can we move this into its own function? `add_accessors_for_field` or similar?

Reply:
Done